### ISSUE TITLE
`Programs` should also call `Program`

### ISSUE THREAD

**Issue by yegor256** [objectionary/lints#123]
Currently, we have two entry points to the library: `Program` to be called in order to check one XMIR file, and `Programs` to be called to check many XMIR files. It's not convenient for the user to use both. Let's make `Program` private and allow the user to use only the `Programs`. 

Also, `Programs` should call `Program`, to make sure that the lints that check individual files are executed.

Don't forget to update the `README.md`.

**Comment by catdog905** [comment]
@yegor256 Would you mind considering Program **interface** introduction. It looks weird to implement logic, you described, in Programs class. Seems like Program class and Programs class have same behavior. They return defects of some software with ability to disable some linters. If both of these classes will implement some interface, we will able to write separate class that will be responsible for providing right implementation based on input argument. Just like you asked in issue description. 

```
public interface Program<T> {
    Collection<Defect> defects();

    Program<T> without;
}
```

```
public final class ProgramOf<T> {
    
    private final Program<T> origin;

    @Override
    public Collection<Defect> defects() {
        this.origin.defects();
    }

    ProgramOf(final Program<T> origin) {
        this.origin = origin;
    }
    
    public Program(final XML xml, final Iterable<Lint<XML>> list) {
        this(
            new SingleProgram<>(
               xml,
               list
            )
         )
    }

    public Program(final Map<String, XML> map, final Iterable<Lint<Map<String, XML>>> list) {
        this(
            new MultipleFileProgram<>(
               map,
               list
            )
         )
    }
}
```

where `SingleProgram` and `MultipleFileProgram`, are current Program and Programs respectively.

**Comment by yegor256** [comment]
@h1alexbel @maxonfjvipon what's your take on this?

**Comment by h1alexbel** [comment]
> [@yegor256](https://github.com/yegor256) Would you mind considering Program **interface** introduction. It looks weird to implement logic, you described, in Programs class. Seems like Program class and Programs class have same behavior. They return defects of some software with ability to disable some linters. If both of these classes will implement some interface, we will able to write separate class that will be responsible for providing right implementation based on input argument. Just like you asked in issue description.
> 
> ```
> public interface Program<T> {
>     Collection<Defect> defects();
> 
>     Program<T> without(String... names);
> }
> ```
> 
> ```
> public final class ProgramOf<T> {
>     
>     private final Program<T> origin;
> 
>     @Override
>     public Collection<Defect> defects() {
>         this.origin.defects();
>     }
> 
>     @Override
>     public Program<T> without(final String... names) {
>         this.origin.without();
>     }
> 
>     ProgramOf(final Program<T> origin) {
>         this.origin = origin;
>     }
>     
>     public Program(final XML xml, final Iterable<Lint<XML>> list) {
>         this(
>             new SingleProgram<>(
>                xml,
>                list
>             )
>          )
>     }
> 
>     public Program(final Map<String, XML> map, final Iterable<Lint<Map<String, XML>>> list) {
>         this(
>             new MultipleFileProgram<>(
>                map,
>                list
>             )
>          )
>     }
> }
> ```
> 
> where `SingleProgram` and `MultipleFileProgram`, are current Program and Programs respectively.

@catdog905 I like this, but what will be the entry point to the `lints` library in this case? If I correctly understood your proposal, in order to lint our XMIR files in both scopes: single program and WPA, we will need to create both `Program` instances:

```java
final Program single = ...;
final Program wpa = ...;
single.defects(); // Single program defects
wpa.defects(); // WPA defects
```

I believe, that more convenient way is to instantiate and lint both scopes just once:

```java
final Program program = ...;
program.defects(); // finds defects in both scopes: Single and WPA
```

WDYT?

**Comment by catdog905** [comment]
@h1alexbel As I can see, if you are required to lint 2 programs, you have to create 2 Program(s) objects anyway.

> 
final Program program = ...;
program.defects(); // finds defects in both scopes: Single and WPA

This piece of code can find defects of only one Program, and it's does not matter: Single or WPA. 
So again, if you need to lint 2 programs, you will always create 2 different programs.

>
final Program single = ...;
final Program wpa = ...;
single.defects(); // Single program defects
wpa.defects(); // WPA defects

But there is no problem for user until he can create both these objects using one class: `ProgramOf`

>
final Program<XML> single = new ProgramOf<>(singleXml, linters);
final Program<Map<String, XML>> wpa = new ProgramOf<>(wpaXmls, linters);
single.defects(); // Single program defects
wpa.defects(); // WPA defects

Here I assumed that one program can not be single and wpa at once

**Comment by catdog905** [comment]
@h1alexbel Important to note, that I don't really understand differences between single and WPA scopes. And if one software project in EO can be simultaneously linted in 2 scopes, we should introduce a class where defects() calls both single.defects() and wpa.defects(), merges results somehow (Different types: XML and Map<String,  XML>), and returns.

So, first, let's clarify task description. What is really required?

**Comment by h1alexbel** [comment]
@catdog905 the goal of this ticket is to have _only one_ entry point to the `lints` library. Currently, we have both: `Program` and `Programs`. It's not convenient to create both to lint XMIR files in both scopes: single program scope and WPA. Single program scope lints each XMIR file in the isolation. While, WPA scope checks all XMIR files as package (finds defects that can be spotted only when looking into multiple XMIR files, not one). We can achieve desired state by making `Program` as part of `Programs`, so each individual XMIR file in the package will be linted in the single program scope as well. Makes sense?

**Comment by catdog905** [comment]
@h1alexbel Got it. You need a single object that can run both scenarios: linting each XMIR file in isolation and linting the whole project.

```
public interface Project {
    Collection<Defect> singleScopeDefects();

    Collection<Defect> wpaScopeDefects();
}
```

Types of these methods are the same, but I believe, they should not be mixed as semantically these two methods return different things

```
public final class ProjectOf implements Project {

    private final Iterable<Program> programs;

    private final Programs wpa;

    @Override
    public Collection<Defect> singleScopeDefects() {
        return new Joined(
            new Mapped(
                Program::defects,
                programs
            )
        );
    }

    @Override
    public Collection<Defect> wpaScopeDefects() {
         return wpa.defects();
    }
    
    ProjectOf(final Iterable<Program> programs, final Programs wpa) {
        this.program = program;
        this.wpa = wpa;
    }

    public ProjectOf(final Map<String, XML> map, final Iterable<Lint<XML>> single, final Iterable<Lint<Map<String, XML>>> wpa) {
        this(
            new Mapped(
                xml -> new Program(xml, single),
                map.values()
            ),
            new Programs(map, wpa)
        );
    }

    // Other constructors to load project from Path, etc
}
```

You can be tempted to put all this logic to existing Programs class, but I believe provided by me design is better as it prevent Programs class to became too complex. Implementing it in a separate class will simplify unit testing and benchmark execution. Overall, I think several small classes is better that big.

@yegor256 @h1alexbel What are your opinions?

**Comment by h1alexbel** [comment]
I like the idea of `Project`, but why we need two methods? I believe it is enough have just one method: `defects()`, that will return all defect. WDYT?

**Comment by catdog905** [comment]
@h1alexbel I'm not sure, but I guess there can be a scenario when user wants to get either single or wpa defects, but not both. Still, I can make one method instead of two if it is better.

**Comment by catdog905** [comment]
@yegor256 @h1alexbel Hi again! Is everything okay with my proposal? Can I start working on this issue?

**Comment by yegor256** [comment]
@catdog905 what exactly is the proposal now? Can you formulate it again, after the discussion has happened above.

**Comment by h1alexbel** [comment]
@Marat-Tim most probably, and we should be able to make `Source` package-private, i.e. closed for users of the library

**Comment by h1alexbel** [comment]
@Marat-Tim the original ticket was about making `Program` execute `Source` (former Programs and Program respectively) internally, and expose only `Program` class for end users. Instead of having two entry points: `Source` and `Program`, the user will have just one: `new Program(xmirs...).defects()`

**Comment by VasilevNStas** [comment]
@yegor256 @volodya-lombrozo @maxonfjvipon, I believe this issue is no longer relevant: the `Program` class (and the whole whole-program-analysis API) was **removed** in commit `235cda2d` (`#845: Refactor XMIR analysis classes for simplicity`, Apr 24, 2026). The entry points today are `Source` and the mono-scope lints (`MonoLints`/`PkMono`); there is no `Program`/`Programs` pair to unify anymore. The README was already updated by that refactoring.

Since both classes mentioned here are gone, this task has no code left to act on. I'd suggest closing it as resolved-by-refactoring. If the WPA-style API is ever reintroduced, a single entry point should be kept by design.

**Comment by volodya-lombrozo** [comment]
@VasilevNStas Yes, indeed. We don't have `Program`s anymore. Lints that touch several files at once were moved into wpa repository:
https://github.com/objectionary/wpa
